staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go

```diff
@@ -682,3 +683,83 @@ func TestGracefulShutdown(t *testing.T) {
 		t.Errorf("Timed out waiting for response.")
 	}
 }
+
+func TestWarningWithRequestTimeout(t *testing.T) {
+	type result struct {
+		err        interface{}
+		stackTrace string
+	}
+	waitForeverCh := make(chan struct{})
```

Comment:
rather than an out-of-band channel like this, I think we can make the test wait on the request context Done() channel

it would also be good to write headers *while* we're waiting for the request to time out to exercise the read / write race, not just the "Header after Handler finished" case

something like this:

```diff
diff --git a/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go b/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go
index 640daaf9ce6..93cf7579072 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/genericapiserver_test.go
@@ -689,13 +689,8 @@ func TestWarningWithRequestTimeout(t *testing.T) {
 		err        interface{}
 		stackTrace string
 	}
-	waitForeverCh := make(chan struct{})
 	resultCh := make(chan result, 1)
 	testHandler := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		// let this request timeout first, the test should close this
-		// channel only after it receives a response for the request.
-		<-waitForeverCh
-
 		defer func() {
 			result := result{}
 			result.err = recover()
@@ -710,9 +705,23 @@ func TestWarningWithRequestTimeout(t *testing.T) {
 			resultCh <- result
 		}()
 
-		// the request has timed out, net/http has finished the handler and
-		// returned a response to the user. Try writing to the response header,
-		// we expect it to not panic.
+		// add warnings while we're waiting for the request to timeout to catch read/write races
+	loop:
+		for {
+			select {
+			case <-r.Context().Done():
+				break loop
+			default:
+				warning.AddWarning(r.Context(), "a", "1")
+			}
+		}
+
+		// the request has just timed out, write to catch read/write races
+		warning.AddWarning(r.Context(), "agent", "text")
+
+		// give time for the timeout response to be written, then try to
+		// write a response header to catch "Header after Handler finished" panics
+		time.Sleep(time.Second)
 		warning.AddWarning(r.Context(), "agent", "text")
 	})
 	handler := newGenericAPIServerHandlerChain(t, "/test", testHandler)
@@ -736,10 +745,6 @@ func TestWarningWithRequestTimeout(t *testing.T) {
 		t.Errorf("expected server to return %d, but got: %v", want, response)
 	}
 
-	// the request has timed out, now we can let the waiting handler
-	// add a warning to the response header.
-	close(waitForeverCh)
-
 	var resultGot result
 	select {
 	case resultGot = <-resultCh:
```